src/flag_gems/ops/pad.py

```diff
@@ -0,0 +1,455 @@
+) -> IndentedBuffer:
+    # make the inlined function visible in the context
+    code.newline()
+
+    # the decorators
+    code.writeline("@libentry()")
+    non_specialize_arg_names = ["value"]
+    code.writeline(f"@triton.jit(do_not_specialize={non_specialize_arg_names})")
+
+    # signature
+    code.writeline(f"def {kernel_name}(")
+    function_ns = NameSpace()
+    with code.indent():
+        code.writeline("in0_ptr: tl.tensor, # of tl.pointer_type")
+        function_ns.create_name("in0_ptr")
+
+        code.writeline("out0_ptr: tl.tensor, # of tl.pointer_type")
+        function_ns.create_name("out0_ptr")
+
+        if rank > 0:
+            # shape for inputs
+            for j in range(rank):
+                function_ns.create_name(f"x_shape{j}")
+            shape_args = ", ".join(f"x_shape{j}: int" for j in range(rank))
+            code.writeline(f"{shape_args}, # shape for x")
+
+            # shape for inputs
+            for j in range(rank):
+                function_ns.create_name(f"in_strides{j}")
+            stride_args = ", ".join(f"in_strides{j}: int" for j in range(rank))
+            code.writeline(f"{stride_args}, # stride for x")
+
+            # shape for inputs
+            for j in range(rank):
+                function_ns.create_name(f"out_strides{j}")
+            stride_args = ", ".join(f"out_strides{j}: int" for j in range(rank))
+            code.writeline(f"{stride_args}, # stride for out")
+
+            # shape for inputs
+            for j in range(rank):
+                function_ns.create_name(f"valid_dim{j}_start")
+            stride_args = ", ".join(f"valid_dim{j}_start: int" for j in range(rank))
+            code.writeline(f"{stride_args}, # valid dim start")
+
+            # shape for inputs
+            for j in range(rank):
+                function_ns.create_name(f"valid_dim{j}_end")
+            stride_args = ", ".join(f"valid_dim{j}_end: int" for j in range(rank))
+            code.writeline(f"{stride_args}, # valid dim end")
+
+            code.writeline("in_elem_cnt: tl.constexpr, ")
+            code.writeline("out_elem_cnt: tl.constexpr, ")
+            code.writeline("value, # padding value")
+            code.writeline("IS_CONSTANT, ")
+            code.writeline("IS_REFLECT, ")
+            code.writeline("IS_REPLICATE, ")
+            code.writeline("IS_CIRCULAR, ")
+            code.writeline("BLOCK_SIZE: tl.constexpr, ")
+
+    code.writeline("):")
+
+    with code.indent():
+        code.writeline("pid = tl.program_id(0)")
+        code.writeline("block_offset = pid * BLOCK_SIZE")
+        code.writeline("offset = block_offset + tl.arange(0, BLOCK_SIZE)")
+        code.newline()
+
+        code.writeline("remaining = offset ")
+        for i in range(rank):
+            code.writeline(f"idx = remaining // out_strides{i}")
+            code.writeline(f"dst_index_{i} = idx")
+            code.writeline(f"remaining = remaining - idx * out_strides{i}")
+            code.newline()
```

Comment:
This part to calculate the multi-index on tensor out is not correct.

since `src_index`s are computed from `dst_index`s, where these indices are logical multi-indx for a multi-dimensional tensor. We can also see that when computing memory offsets on tensor in with following code:

```python
src_offset = src_index_0 * in_strides0
src_offset += src_index_1 * in_strides1
```

Since we are unraveling `offset` to get the logical multi-index, it is the tensor shape that should be considered(in divide and modular operations).

A simple test case where the input is not C-contiguous can verify this.

```python
@pytest.mark.parametrize("shape", [(100, 100)])
@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
def test_constant_pad(shape, dtype):
    x = torch.randn(size=shape, dtype=dtype, device="cuda")[::2, ::2]
    ref_x = to_reference(x)

    rank = x.ndim
    pad_params = tuple(np.random.randint(0, 10, rank * 2))

    pad_value = np.random.randint(0, 1024, 1)[0]
    ref_out = torch.nn.functional.pad(x, pad_params, "constant", pad_value)
    with flag_gems.use_gems():
        res_out = torch.nn.functional.pad(ref_x, pad_params, "constant", pad_value)

    gems_assert_equal(ref_out, res_out)
```

Reply:
good catch, fixed it by change `load_cond` to check index is valid for input shape instead of offset